zerver/lib/actions.py

```diff
@@ -1284,6 +1285,93 @@ def do_delete_user(user_profile: UserProfile) -> None:
         )
 
 
+def do_delete_user_preserving_messages(user_profile: UserProfile) -> None:
+    """
+    This is a version of do_delete_user which does not delete message content.
+
+    The code is a bit tricky, because we want to, at some point, call
+    user_profile.delete() to trigger cascade deletes to clean up the variety
+    of related objects - but we need to avoid the cascades affecting
+    messages sent by the user.
+    """
+    if user_profile.realm.is_zephyr_mirror_realm:
+        raise AssertionError("Deleting zephyr mirror users is not supported")
+
+    do_deactivate_user(user_profile, acting_user=None)
+
+    subscribed_huddle_recipient_ids = set(
+        Subscription.objects.filter(
+            user_profile=user_profile, recipient__type=Recipient.HUDDLE
+        ).values_list("recipient_id", flat=True)
+    )
+    user_id = user_profile.id
+    personal_recipient = user_profile.recipient
+    realm = user_profile.realm
+    date_joined = user_profile.date_joined
+
+    with transaction.atomic():
+        # The strategy is that before calling user_profile.delete(), we need to
+        # reassign Messages  sent by the user to a dummy user, so that they don't
+        # get affected by CASCADE. We cannot yet create a dummy user with .id
+        # matching that of the user_profile, so the general scheme is:
+        # 1. We create a *temporary* dummy for the initial re-assignment of messages.
+        # 2. We delete the UserProfile.
+        # 3. We create a replacement dummy user with its id matching what the UserProfile had.
+        # 4. This is the intended, final replacement UserProfile, so we re-assign
+        #    the messages from step (1) to it and delete the temporary dummy.
+        random_token = secrets.token_hex(16)
+        temp_replacement_user = create_user(
+            email=f"temp_deleteduser{random_token}@{get_fake_email_domain(realm)}",
+            password=None,
+            realm=realm,
+            full_name=f"Deleted User {random_token}",
+            active=False,
+            is_mirror_dummy=True,
+            force_date_joined=date_joined,
+            create_personal_recipient=False,
+        )
+        Message.objects.filter(sender=user_profile).update(sender=temp_replacement_user)
+        user_profile.delete()
```

Comment:
Does the `delete` method support overriding the `ON_DELETE` behavior for a single operation?  Because it might be easier if we could just preventing `ON_DELETE=CASCASE` from applying.

Reply:
This does have the nice property of being explicit about what it doesn't cascade to -- namely, the things we move out of the way, and the huddle subscriptions that we re-create.  Switching the cascade behaviour may leave lingering other pieces of data not deleted.  For example, I think this code deletes RealmEmoji that the user made, because we don't preserve them explicitly.  I'm not sure if that's right or wrong, but it seems more expected from the code as written now.

Reply:
Yeah, the idea here was **wanting** to have the CASCADE behavior to know we automatically delete anything that's not explicitly meant to be kept. Otherwise we could even do a dirty `DELETE` direct SQL query to delete the row, because the CASCADE is enforced at the Django level rather the Postgres level, so it can easily be bypassed that way.

Reply:
Yeah, I think this is a defensible strategy. The main side effect is spending a user ID in the UserProfile sequence, which is wasteful but probably OK as long as we document the choice clearly.